 1 init.rc: mix device-specific data into the Linux RNG
Mix the contents of /proc/cmdline and /default.prop
into /dev/urandom. /proc/cmdline often contains
androidboot.serialno, a device-specific unique
identifier. Similarly, /default.prop contains the
build fingerprint and timestamp, which vary between
device families.
 2 init.rc: add missing /dev/fd symlink
The Linux kernel implicitly expects /dev/fd to symlink to /proc/self/fd.
This change fixes the exec/execveat.c kernel selftest.
 3 init.rc: restart keystore when servicemanager restarts

diff --git a/rootdir/init.rc b/rootdir/init.rc
index 030d248..13fc5fa 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -29,6 +29,10 @@ on early-init
 on init
     sysclktz 0
 
+    # Mix device-specific information into the entropy pool
+    copy /proc/cmdline /dev/urandom
+    copy /default.prop /dev/urandom
+
     # Backward compatibility.
     symlink /system/etc /etc
     symlink /sys/kernel/debug /d
diff --git a/rootdir/init.rc b/rootdir/init.rc
index 030d258..13fc5fa 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -216,6 +216,10 @@ on init
     # enable armv8_deprecated instruction hooks
     write /proc/sys/abi/swp 1
 
+    # Linux's execveat() syscall may construct paths containing /dev/fd
+    # expecting it to point to /proc/self/fd
+    symlink /proc/self/fd /dev/fd
+
 # Healthd can trigger a full boot from charger mode by signaling this
 # property when the power button is held.
 on property:sys.boot_from_charger_mode=1
diff --git a/rootdir/init.rc b/rootdir/init.rc
index 030d268..13fc5fa 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -633,6 +633,7 @@ service servicemanager /system/bin/servicemanager
     onrestart restart media
     onrestart restart surfaceflinger
     onrestart restart drm
+    onrestart restart keystore
 
 service vold /system/bin/vold \
         --blkid_context=u:r:blkid:s0 --blkid_untrusted_context=u:r:blkid_untrusted:s0 \
